### PR TITLE
Allow copying a ParameterizedTypeName with new type arguments

### LINKED ISSUE #1139: Support copying a ParameterizedTypeName with different type arguments
I need to create a wildcard projected type name and currently there is no way to do so using public APIs. I suggest a `copy` function is added which has all of the parameters of the class.

### PR DESCRIPTION
**Pull request by ansman**
This fixes #1139

### CODE REVIEW

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/ParameterizedTypeNameTest.kt [hunk 1]
```diff
@@ -112,6 +112,15 @@ class ParameterizedTypeNameTest {
     assertThat(typeName.toString()).isEqualTo("java.util.Map<java.lang.String, java.lang.Integer>")
   }
 
+  @Test fun copyingTypeArguments() {
+    java.util.Map::class.java.plusParameter(java.lang.String::class.java)
```
Comment (Egorand) [inline kotlinpoet/src/test/java/com/squareup/kotlinpoet/ParameterizedTypeNameTest.kt:116]: Does this line do anything?

**Review by Egorand — APPROVED** [review]
Thanks!

File: kotlinpoet/src/test/java/com/squareup/kotlinpoet/ParameterizedTypeNameTest.kt (hunk 1 above)
Comment (ansman) [reply kotlinpoet/src/test/java/com/squareup/kotlinpoet/ParameterizedTypeNameTest.kt:116]: Damn, I left some debug stuff. But we can pretend I left it there to check if you would notice.

Comment (Egorand) [reply kotlinpoet/src/test/java/com/squareup/kotlinpoet/ParameterizedTypeNameTest.kt]: I passed!

**Comment by ZacSweers** [comment]
From an API standpoint I wonder if it would be better to have a withParameters instead. These overloaded copy functions are brittle historically

**Comment by ansman** [comment]
I'd be OK with that too. In fact I was probably going to add a `mapTypeArguments` in my project too.

**Comment by Egorand** [comment]
IMO `copy` is easier to discover with it being a Kotlin pattern. What's causing brittleness, the fact that it's hard to change in a backwards-compatible way?

**Comment by ansman** [comment]
If we are keeping copy I think it might make sense to expose the enclosing type and add both that and the raw type to the copy method.

**Comment by Egorand** [comment]
@ansman I'd be reluctant to expose those simply for consistency, without a use case. If there is a use case then we'll consider it.

**Comment by ansman** [comment]
My current use cases are:
* Creating a wildcarded typename.
* Mapping Java types to Kotlin types (`List`, `Set`, `Annotation` etc). 

This isn't possible in all cases without exposing these things.

**Comment by ZacSweers** [comment]
Can you add a test case that wasn't possible before? The one here is possible with the previous API

**Comment by ansman** [comment]
Done

**Review by ZacSweers — APPROVED** [review]
This seems ok to me, though that example has me 🤔🤔🤔

**Comment by ansman** [comment]
Yeah, the example isn't perhaps real since you'd access this as `Map.Entry<...>`. I essentially needed a nested class and the enclosing class needs to be parameterized to test the specific case.

**Pull request merged**